Add owning user and group, together with unix permissions

believeinlain opened the pull request:
Addresses issue https://github.com/sourcefrog/conserve/issues/46

- Added file unix_mode.rs that adds metadata representing the Unix mode bits using std::fs::Permissions and std::os::unix::fs::PermissionsExt (on unix).
- Added display of the form "rwxrwxrwx" (with set uid, set gid, and sticky bits) to backup, restore, and ls commands.
- Still some work to do adding and testing Windows compatibility (the plan is to always store archives with Unix mode bits, but to use a systematic conversion when storing/retrieving files on Windows).
- Works with old archives by setting the default mode bits to 0o100775, which should ideally be changed, as a more suitable default is 0o664 for files and 0o775 for directories. (I think the 1 in 0o100775 is the inode type, as in file, directory, symlink, etc., but I'm not sure). Regardless, it captures the entire unix mode so it works well on unix-like systems with newer backups. It should in the future be extended to set an appropriate default for better compatibility with other OSs and old archives.
- Added owner.rs which stores the user and group owning each entry as a String.
- Added display of the form "user group" to backup, restore, and ls commands.
- Developed test "backup_user_and_permissions" to ensure that user, group, and permissions are saved and restored properly on Unix.

sourcefrog commented:
Thanks @believeinlain! I'll have a look when I have some time.

sourcefrog commented on src/entry.rs:
```diff
@@ -26,12 +28,15 @@ pub trait Entry: Debug + Eq + PartialEq {
         basis_entry.kind() == self.kind()
             && basis_entry.mtime() == self.mtime()
             && basis_entry.size() == self.size()
+            && basis_entry.umode() == self.umode()
```
Seems like this should also check the owner?

believeinlain replied:
Yes I think so

sourcefrog commented on src/unix_mode.rs:
```diff
@@ -0,0 +1,172 @@
+use std::{fmt, fs::Permissions};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct UnixMode {
```
I actually have an existing crate that implements dissecting and parsing similar to this. I'd rather add a dependency: https://github.com/sourcefrog/unix_mode, https://crates.io/crates/unix_mode

believeinlain replied:
Nice, I'll use that instead

sourcefrog commented on src/unix_mode.rs:
```diff
@@ -0,0 +1,172 @@
+    fn default() -> Self {
+        // created with execute permission so that restoring old archives works properly
+        // TODO: ideally we would set this based on the inode type read from the archive
+        Self { mode: 0o100775 }
```
I think we should probably strip off the top file-type bits and store only the permissions?

believeinlain replied:
Agreed. IIRC I had those bits in there to get the old archive test to pass, since for the most part the mode is set from reading the file metadata rather than relying on the default, but when reading archives from the testdata there were some problems.
Now that I think though it could also be accomplished with a mask in the Eq implementation rather than comparing all the bits in mode, which would be cleaner.

believeinlain replied:
Current solution is to store the inode type bits if provided but not compare them in the Eq impl, and not print them in the display impl unless they are present. I considered just stripping off the first character of the display string no matter what (as I had it previously) but I don't see how it hurts to print the inode type if the bits happen to be present.

believeinlain replied:
Regardless, it won't be hard to make further changes.

sourcefrog replied:
The type bits are redundant with the separately stored file kind so I think we should strip them are the point of reading and only store the permissions in this field.

believeinlain replied:
Ok, you do have a point there.

sourcefrog commented on src/index.rs:
```diff
@@ -49,6 +51,14 @@ pub struct IndexEntry {
     #[serde(default)]
     pub mtime: i64,
 
+    /// Discretionary Access Control permissions (such as read/write/execute on unix)
```
Could you please also mention these in https://github.com/sourcefrog/conserve/blob/main/doc/format.md

believeinlain replied:
done

sourcefrog commented on src/index.rs:
```diff
@@ -49,6 +51,14 @@ pub struct IndexEntry {
 
+    /// Discretionary Access Control permissions (such as read/write/execute on unix)
+    #[serde(default)]
+    pub umode: UnixMode,
```
I like the style rule of making the variable name match the type name, so could you please `s/umode/unix_mode/`? (Also on the function names and so on.)

believeinlain replied:
Done

sourcefrog commented on src/index.rs:
```diff
@@ -523,14 +545,18 @@ mod tests {
              \"kind\":\"File\",\
-             \"mtime\":1461736377}]"
+             \"mtime\":1461736377,\
+             \"umode\":{\"mode\":33277},\
```
It seems like we don't really need another layer of dict here, because the mode is really unlikely to have more than a u32... We could either:

1. Make `struct UnixMode(u32)`
2. Or tell serde to flatten it (https://serde.rs/attr-flatten.html)

I think I prefer 1...

believeinlain replied:
Cool, I didn't think of that. Changed UnixMode to a tuple struct: `UnixMode(u32)`

sourcefrog commented on src/index.rs:
```diff
@@ -523,14 +545,18 @@ mod tests {
-             \"mtime\":1461736377}]"
+             \"mtime\":1461736377,\
+             \"umode\":{\"mode\":33277},\
+             \"owner\":{\"user\":null,\"group\":null}}]"
```
I also wonder if these should be flattened into the index entry, https://serde.rs/attr-flatten.html.

sourcefrog replied:
Also, could we omit them if they're null? It seems a bit wasteful to serialize.

believeinlain replied:
Hmm. There seem to be a few options of how to fix this. 
How likely do you think it would be to have the owning user defined but not the group or vice versa? If we store an Option<Owner> rather than an Owner { Option, Option } that would solve it, but then either both user and group need to be defined or neither.
Or I could just impl the serialize and deserialize functions manually so it's more compact if null, but that would make the archive difficult to read using other tools - I think it's best to stick with standard serde serialization.
Then there's flattening, and I imagine serde has a way to simply not include null valuesin the serialization - I think flattening is probably the best way to go.

believeinlain replied:
I decided to set flatten the Owner struct and `skip_serializing_if` both the user and group are None, so the null values won't get stored unless only one of the user or group are defined, which seems unlikely.

sourcefrog replied:
I'm not really sure what would be helpful on Windows, but from memory it doesn't have a strongly distinguished file group the way Unix does. So we might want to store only the owner there...

believeinlain replied:
I don't really have a Windows development environment at the moment so I'm not going to be able to test and develop Windows functionality for a while.

sourcefrog commented on src/show.rs:
```diff
@@ -117,10 +117,18 @@ pub fn show_index_json(band: &Band, w: &mut dyn Write) -> Result<()> {
     for entry in it {
-        writeln!(bw, "{}", entry.apath())?;
+        if long_listing {
+            writeln!(bw, "{} {} {}", entry.umode(), entry.owner(), entry.apath())?;
```
It seems like this will get messy if the owners vary from one line to another since it's not doing any column alignment...

believeinlain replied:
I'm not sure how best to address this. I agree it could get messy but I'm not sure it's worth doing a proper column alignment.

sourcefrog replied:
In a sense the best approach, and what's probably done by table-drawing crates, is to find the maximum length for each column and then use that to set the width. However that would mean reading the whole index into memory before printing anything, which goes against a general rule for Conserve of avoiding ever needing to have the whole index in memory.

So pragmatically I think just setting them to a fixed width of 10 for the mode and 10(??) for the user and group would be fine.

(I think modern Unix has a long limit on usernames, maybe 32 characters, which would make this unreasonably wide. But under 10 seems very common?)

So just `{:<10}` (https://doc.rust-lang.org/std/fmt/#fillalignment)

believeinlain replied:
Ok cool, that's an easy change.

sourcefrog commented on tests/cli/main.rs:
```diff
@@ -274,6 +274,119 @@ fn basic_backup() {
 
+#[test]
+#[cfg(unix)]
+fn backup_user_and_permissions() {
```
It would be nice to make this test stronger. I suspect it might almost pass even if the owners and perms are not stored...

Testing changes to the owner and group may be difficult in a hermetic way. Especially if the tests aren't running as root we probably cannot set the user.

We could look at what groups the test runner belongs to, and change some files to be owned by different groups? There is very likely more than one? (Maybe not in CI?)

For permissions, it would be nice to make some files executable, readonly, and group-writeable, and check that the permissions are restored.

In particular there could conceivably be a bug where a restored file is made readonly too soon and so can't be written. It'd be nice to have a test showing that readonly files are properly restored.

believeinlain replied:
Agreed, this test should be more robust.

believeinlain replied:
Added the test `backup_unix_permissions` which should properly test the unix permissions. Unsure how to test the user/group so I'm just gonna hold off on that for now.

sourcefrog replied:
I had a look at this and extended it a bit, and I think it's OK for now. More thoroughly testing this probably requires some scaffolding to save indexes that don't exist on disk. That'd be good to have but it's well beyond the scope of this PR.

sourcefrog commented on src/restore.rs:
```diff
@@ -226,6 +251,31 @@ impl RestoreTree {
+                None
+            };
+            // TODO: use `std::os::unix::fs::chown(path, uid, gid)?;` once stable
+            unistd::chown(
```
Maybe this should not be fatal? It seems like restoring on a different machine could easily fail to restore users/perms but it should not prevent getting the data back.

believeinlain replied:
Good point.

sourcefrog commented on tests/cli/main.rs:
```diff
@@ -274,6 +274,119 @@ fn basic_backup() {
     // TODO: Compare vs source tree.
 }
```
There are also some tests for "old formats" to check we can read archives written by old versions, from `testdata/archive`. When this is finalized we should check in a small archive there.

It would also be good to check `ls -l` on an old archive. I think it will just work because it's stored in Options.

believeinlain replied:
The mode and owner should be set to the default if they aren't found in the archive. Currently I think that means the owner would be "", which doesn't make sense. I think it should probably be set to the user running conserve if it can't be found in the archive, since that's the user that will own the files upon restoration.

believeinlain replied:
I made a test `long_listing_old_archive` that calls `show::show_entry_names` on the `MINIMAL_ARCHIVE_VERSIONS`. Metadata just shows up as `rwxrwxr-x none none` because with no data stored it just assumes the default values. 
An improvement would be to set the default metadata more intelligently (based on current user/group and inode type) but I don't think it's worth the trouble.

sourcefrog replied:
That seems OK for listing. It might be nicer to show the permissions also as `none` though.

It does make me think that we should test that _restore_ from an archive that has no permissions or owners should not chmod or chown the files.

believeinlain replied:
We could have the unix_mode be stored as an `Option<u32>` so that the default is `None`. Testing that it doesn't chown or chmod the files would be tricky, since the files typically don't exist prior to restoring them from the archive. Could be done by invoking API functions, but I think if the mode and owner are stored as `None` that's reasonable enough assurance that nothing will be overridden.

sourcefrog reviewed (commented):
Thanks, that looks good!

sourcefrog commented on src/restore.rs:
```diff
@@ -104,7 +112,16 @@ pub fn restore(
     )?;
     for entry in entry_iter {
         if options.print_filenames {
-            progress_bar.message(&format!("{}\n", entry.apath()));
```
Ideally I should unify the things that print like this, but this PR is already big, so not here.

sourcefrog commented on src/index.rs:
```diff
@@ -49,6 +51,14 @@ pub struct IndexEntry {
 
+    /// Discretionary Access Control permissions (such as read/write/execute on unix)
+    #[serde(default)]
+    pub unix_mode: UnixMode,
```
Actually, thinking about the case of old formats and backups from Windows: it seems like these should be `Option` rather than set to the default if missing. Then, restore should only attempt to set the permissions/owner if they are set.

believeinlain replied:
UnixMode is now UnixMode(Option<u32>), and restore cannot invoke fs::set_permissions if UnixMode is None, because it can't extract a Permissions struct from None.

believeinlain commented:
Ok, I believe I've addressed your comments with exception of a comprehensive Windows implementation of permissions and owner and a more robust test for owning user and group saving/restoring. I probably won't be able to get to either of those for a while, so I leave it to you to decide how you want to move forward with this PR.

sourcefrog commented:
Hi, thanks. This is already pretty big and a pretty good feature by itself, so I don't think we should necessarily block on adding more Windows support. We might want options to not restore ownership and permissions or maybe not even store it, but that also could be done later.

I read the later code changes and they look fine. I will try out the code and have another look at the tests some time soon.

sourcefrog commented:
It looks like the tests are a bit brittle, probably depending on the umask where they're run.

sourcefrog commented:
OK, sorry for the delay. I think this is nice. A few more things I'll finish up and then merge it:

- [ ] Reconcile with small changes in `main`
- [ ] Save a binary copy of a tree with users and permissions into `testdata` so that we can check future versions can read it; and actually test against it.
- [ ] Test restoring groups, assuming that the user running the tests is in more than one group. (Almost always true locally; might not be true in Github CI; but we could make the test skip.)
- [ ] Maybe there should be an option to restore without restoring user and group. In particular without this it might be impossible to restore from canned data containing users and groups... Or, maybe it's simpler to just try it and emit warnings if the required user and group don't exist...

sourcefrog commented:
OK it's pretty clear from the code that attempting to restore a user or group who doesn't exist locally won't fail, so I'll leave that as is for now.

The Clippy checks from CI are a bit annoying, especially as Clippy may get more pedantic over time independently of any changes in the code. I'll turn them off.
